EDUCATOR-434: Unit Group Access

jlajoie opened the pull request:
## [EDUCATOR-434](https://openedx.atlassian.net/browse/EDUCATOR-434)

### Description

Allows a user to configure Unit group access from the course outline page. Adds tests, additional warning messages, and separates group access code from staff lock code.

### Sandbox
- [x] [https://studio-outline-group-access.sandbox.edx.org](https://studio-outline-group-access.sandbox.edx.org)
- [x] [Direct link to course outline](https://studio-outline-group-access.sandbox.edx.org/course/course-v1:edX+DemoX+Demo_Course?show=block-v1%3AedX%2BDemoX%2BDemo_Course%2Btype%40sequential%2Bblock%40edx_introduction)

1. Access the demo course in studio
2. Create a Unit
3. On the Course Outline page select the gear for a Unit, and set either the enrollment track or content group options.
4. Verify content warning messages are correct on Course Outline and that access for users exists on LMS

### Testing
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @cahrens 
- [x] Code review: @ssemenova 
- [ ] Doc Review: @catong 

FYI: @staubina @bbaker6225 @sstack22

cahrens commented on cms/templates/js/staff-lock-editor.underscore:
```diff
@@ -29,4 +29,46 @@
+            </select>
+            </li>
+            <br>
+            <% if (contentGroups.length > 0) { %>
```
Note that you will have to deal with deleted groups, in the same way that visibility_editor.html and its associated code does.

Also don't forget i18n and safe template practices. 
http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/internationalization/i18n.html

http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -591,20 +580,69 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+            }
+        },
+
+        getSelectedEnrollmentTrackId: function() {
```
Can you guys pull in VisibilityEditorView from /edx-platform/cms/static/js/xblock/authoring.js (or refactor it so that the guts can be shared between these 2 JS classes?

jlajoie commented:
@catong This is the WIP PR for unit level access

cahrens commented on cms/templates/js/staff-lock-editor.underscore:
```diff
@@ -29,4 +29,46 @@
         </li>
     </ul>
 </div>
+<br>
```
Note that this file should not remain "staff-lock-editor.underscore".

bbaker6225 commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -190,6 +195,7 @@ def _build_fixture(self, unit_state):
         """ Returns an XBlockFixtureDesc with a section, subsection, and possibly unit that has the given state. """
         name = unit_state.name
         start = (datetime(1984, 3, 4) if unit_state.is_released else datetime.now(UTC) + timedelta(1)).isoformat()
+        id_base = MINIMUM_STATIC_PARTITION_ID
```
Looks like we can get rid of this line now

cahrens commented:
This unit-level access message should not be showing on Subsections.

![image](https://user-images.githubusercontent.com/484484/27234096-d8e9005a-5289-11e7-8e9b-1d1a4efe5718.png)

cahrens commented:
I see the checkmark styling is different from what we have on the unit page dialog. But looking at stage, it is consistent with what was there before. I can't say I'm loving it though. Please do include @roderickmorales as a reviewer.

![image](https://user-images.githubusercontent.com/484484/27234142-f2a27184-5289-11e7-9d90-da1d93298c75.png)

Also, is it possible to make the default dialog size larger so that it doesn't have to resize to show 2 enrollment tracks (the most common scenario)? I'm not sure this is possible giving the dynamic nature of how the dialog is constructed.

cahrens commented:
The deleted group logic needs to be present in this dialog (as it is on the unit page/component level dialog). It must show when a unit is linked to a group that no longer exists.

Here is a way to get into the state:
1) Create a new Content Group in one web browser tab
2) In a different browser tab, bring up the new unit level access dialog, select restrict access by Content Groups, but don't select anything (leave dialog up).
3) In the first tab, delete the new Content Group (possible because nobody is using it)
4) In the second tab, select the deleted Content Group in the open dialog and save it.

This is the state you get in that dialog. In does not show that the deleted group is selected (nor give you a way to unselect it):

![image](https://user-images.githubusercontent.com/484484/27234557-69fab042-528b-11e7-9da6-be9500e211f9.png)

This is how it should look (note also the help text underneath):

![image](https://user-images.githubusercontent.com/484484/27234658-d102077c-528b-11e7-8728-99c90256d7dd.png)

And note that this is what we get on the Unit page when the unit-level access refers to deleted groups. We are going to need to generalize this language or something....

![image](https://user-images.githubusercontent.com/484484/27234610-97974ad8-528b-11e7-9eed-a7b54a48e4a8.png)

cahrens commented:
I am getting a server error when I try clicking on usage links in Group Configurations that refer to units. For example,

https://studio-outline-group-access.sandbox.edx.org/container/block-v1:edX+DemoX+Demo_Course+type@sequential+block@edx_introduction

Which is the URL (in https://studio-outline-group-access.sandbox.edx.org/group_configurations/course-v1:edX+DemoX+Demo_Course) for 
![image](https://user-images.githubusercontent.com/484484/27234732-05f7a9aa-528c-11e7-96e1-71f8ae584081.png)

It seems like a new story needs to be created for fixing these links @jlajoie

cahrens commented:
@sstack22 and @roderickmorales should both be reviewers as well.

cahrens commented on cms/templates/js/visibility-and-access-editor.underscore:
```diff
@@ -0,0 +1,78 @@
+        <ul class="list-fields list-input">
+            <li>
+            <%- gettext('Restrict access to:') %>
+            <select class="user-partition-select">
```
The select has no label associated with it. This is an a11y violation, as you can see in your browser if you run the accessibility audit tool in Chrome dev tools.

Please look at visibility_editor.html for how you put a label around the select.
Also remove the ul/li structure in here, as it doesn't make sense.

![image](https://user-images.githubusercontent.com/484484/27235215-e9feef9a-528d-11e7-9ecd-c34de81bb893.png)

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -599,6 +589,28 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
         }
     });
 
+    AbstractVisibilityAndAccessEditor = AbstractEditor.extend({
```
Is there any reason to have this abstract class? Isn't it only used one place?

jlajoie replied:
Yes it is only used here, we'll squash it

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -633,6 +645,99 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
 
+    VisibilityAndAccessEditor = AbstractVisibilityAndAccessEditor.extend({
+        templateName: 'visibility-and-access-editor',
+        className: 'edit-staff-lock',
```
Is "edit-staff-lock" really the correct className here?

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -633,6 +645,99 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
     });
 
+    VisibilityAndAccessEditor = AbstractVisibilityAndAccessEditor.extend({
+        templateName: 'visibility-and-access-editor',
```
This template is just the access part now, correct? I think that same renaming/cleanup still needs to be done.

cahrens commented on cms/templates/js/visibility-and-access-editor.underscore:
```diff
@@ -0,0 +1,78 @@
+    })
+})
+%>
+<% if (xblockInfo.isVertical()) { %>
```
This is only used for verticals, correct?

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -798,7 +903,7 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
                 if (xblockInfo.isChapter()) {
                     tabs[0].editors = [ReleaseDateEditor];
-                    tabs[1].editors = [StaffLockEditor];
+                    tabs[1].editors = [StaffLockEditor, VisibilityAndAccessEditor];
```
This shouldn't be shown on Chapters (Subsections).

roderickmorales commented:
I would prefer having the plain checkboxes. The "unchecked" state here looks almost disabled and the I'm the blue "checked" style looks almost like it is too much for what we need

cahrens commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -198,71 +204,128 @@ def _build_fixture(self, unit_state):
                 XBlockFixtureDesc('vertical', name, metadata={
-                    'visible_to_staff_only': True if unit_state.is_locked else None
+                    'visible_to_staff_only': True if unit_state.is_locked else None,
+                    'group_access': {50: [1]} if unit_state.is_restricted else {},
```
I think we have some constants in other tests for these values (see test_studio_container.py).

I don't see that there is test coverage for the dialog itself (either bok choy or sufficient unit test coverage). Can you explain which portions of the code change in this PR are covered by tests? We will want to ensure we have test coverage for using the dialog to control access.

jlajoie replied:
Currently we have the javascript tests for the expected form body and verifying that the divided warning messages are shown, we also have bok choy tests for similar functionality that probably doesn't need to be duplicated. What we can add are bok-choy tests that will go through a user's workflow of creating a group configuration, and then setting the group access on a unit via the course outline page

cahrens commented:
@roderickmorales I agree about the checkboxes (indeed, the unchecked color has a11y issues). @jlajoie and @bbaker6225, when you make this change, it should impact all checkboxes used within these course outline dialogs. Don't try to make it specific to just this access portion.

jlajoie commented:
@cahrens @ssemenova This should be good for another round of reviews, we added additional bok-choy tests and believe we fixed the a11y issues within the modal. Changes for the deleted group messaging on the Course Outline Page we've opened a story for at https://openedx.atlassian.net/browse/EDUCATOR-699 and the bug we introduced for the broken Unit links on the Group Configurations Page we've opened at https://openedx.atlassian.net/browse/EDUCATOR-649

cahrens commented:
I would imagine we shouldn't show anything in the dialog if there are no Content Groups and only a single Enrollment Track. Or perhaps we do want to show a message, but there is no need to have the select. Please work with @roderickmorales about how this should look.

Dialog on Unit Page:
![image](https://user-images.githubusercontent.com/484484/27395560-2d42d516-567e-11e7-9564-5d6af99a4933.png)

Existing dialog on a component in this situation:
![image](https://user-images.githubusercontent.com/484484/27395573-397b41f6-567e-11e7-8364-563591471807.png)

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -242,8 +249,9 @@ define(['jquery', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpers', 'common/j
                     'due-date-editor', 'grading-editor', 'publish-editor',
-                    'staff-lock-editor', 'content-visibility-editor', 'settings-modal-tabs',
-                    'timed-examination-preference-editor', 'access-editor', 'show-correctness-editor'
+                    'staff-lock-editor', 'unit-access-editor', 'content-visibility-editor',
```
Out of curiosity, why is it necessary for this spec to know about all these templates, even though the spec does not seem to be testing the dialog? Not a big deal, just trying to understand.

cahrens replied:
@jlajoie Did you have a chance to look into this? Again, just curious.

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -591,10 +581,20 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
             return this.model.get('ancestor_has_staff_lock');
         },
 
+        getContentGroups: function() {
```
Do these methods need to be in the AbstractVisibilityEditor? In master, that base class is used for StaffLockEditor and ContentVisibilityEditor. I don't think either of those need content group or group access information. See ContentVisibilityEditor for an example of how getContext can be extended in UnitAccessEditor to provide the content group and access information for that specific editor.

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -633,6 +633,103 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+                        // Select the option that has group access, provided there is a specific group within the scheme
+                        this.$('.user-partition-select option[value=' + keys[0] + ']').prop('selected', true);
+                        this.showSelectedDiv(keys[0]);
+                        this.$('#partition-select option:contains("Select a group type")').text(
```
This will not work with internationalization. You can't assume English text or set anything specifically to English text.

Take a look at how the other dialog handles this.

cahrens replied:
I don't see that "Select a group type" ever appears in the other dialog. Can't the text just always be "All Learners and Staff" for the first option in the select?

bbaker6225 replied:
For your second comment, we're copying the UI from the access modal on the component page, where it says "Select a group type" if you haven't previously selected any group to restrict access to.
We previously had it always say "All Learners and Staff" before we changed it for the last review

@roderickmorales Any opinion on which one we should decide on?

cahrens replied:
OK, I now see the condition in which "Select a group type" does appear. However, can you handle this initial state in the template itself?

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -633,6 +633,103 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+                if (groupAccessData) {
+                    metadata.group_access = groupAccessData;
+                }
+                return {
```
To make sure I understand, if the condition at 720 is False, then group access is being cleared (set to All Learners and Staff)?

bbaker6225 replied:
Yes, that's correct

cahrens commented on cms/static/sass/elements/_modal-window.scss:
```diff
@@ -777,4 +780,43 @@
+  .modal-section-content {
+    @include font-size(16);
+
+    .group-select-title {
```
I think the CSS that is specific to the new access editor should be scoped within a class so that it does not accidentally impact other editors. Perhaps put within the edit-unit-access class?

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -4,6 +4,32 @@ var visibilityState = xblockInfo.get('visibility_state');
+    var schemeAllowedGroups = groupAccess[schemeId];
+
+    schemeAllowedGroups.forEach(function (groupId) {
+        var fullId = schemeId + groupId;
```
Isn't this concatenating together 2 integers? If so, that seems like a bad idea because '3' + '10' == '31' + '0', for example.

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -57,11 +83,25 @@ if (staffOnlyMessage) {
+        messageText = edx.StringUtils.interpolate(
+                gettext('Access to this unit is restricted to {groupNames}'),
+                {
+                    groupNames: groupNames.join(', ')
```
This will not work for translation-- not all languages use commas in this way. You need to follow the same pattern as in contentstore/utils.py. Perhaps you can find a way to share the code.


````
# At the same time, build up all the selected groups as they are displayed in the dialog title.
    selected_groups_label = ''

    for index, partition in enumerate(selectable_partitions):
        for group in partition["groups"]:
            if group["selected"]:
                if len(selected_groups_label) == 0:
                    selected_groups_label = group['name']
                else:
                    # Translators: This is building up a list of groups. It is marked for translation because of the
                    # comma, which is used as a separator between each group.
                    selected_groups_label = _('{previous_groups}, {current_group}').format(
                        previous_groups=selected_groups_label,
                        current_group=group['name']
                    )

cahrens commented on cms/templates/js/staff-lock-editor.underscore:
```diff
@@ -7,26 +7,20 @@
-              <span class="icon fa fa-square-o input-checkbox-unchecked" aria-hidden="true"></span>
-              <%- gettext('Hide from learners') %>
-            </label>
+    <input type="checkbox" id="staff_lock" name="staff_lock" class="input input-checkbox" />
```
Nit: if you put the label tag around the input, you don't need to link them together with "for" and "id" values.

cahrens replied:
@roderickmorales Note that we have moved away from "checkbox-cosmetic" in a couple of places, but there will still be other editors in the general dialog that use "checkbox-cosmetic". Hopefully we can eradicate them all during a11y work.

cahrens commented on cms/templates/js/unit-access-editor.underscore:
```diff
@@ -0,0 +1,133 @@
+                trackSelected = true;
+            }
+        });
+        // Only add enrollment tracks if there are multiple tracks to restrict from
```
Or if one of the tracks is already selected.

cahrens commented on cms/templates/js/unit-access-editor.underscore:
```diff
@@ -0,0 +1,133 @@
+    groups.forEach(function (group) {
+        var groupId = group['id'];
+        var groupName = group['name'];
+        var fullId = schemeId.toString() + '-' + groupId.toString();
```
While I am not in love with this concatenation (I find map/dictionary structures much more natural for this sort of hierarchical data), at least you have a dividing character this time.

cahrens commented on cms/templates/js/unit-access-editor.underscore:
```diff
@@ -0,0 +1,133 @@
+    })
+}
+
+// Need to include deleted groups logic to not change is_active boolean in contentstore/utils.py get_user_partition_info
```
Let's not work around the implementation of get_user_partition_info. If a bug exists in it with regard to deleted groups not always showing, the fix should be in that method so that the component access editor on the unit page also picks up the fix.

If this is not trivial to fix, I'm fine with you creating a separate bug (assuming that the behavior you are describing shows up in both dialogs-- if it doesn't show up in both locations, I'd like to understand why).

jlajoie replied:
We can look into how difficult this is to fix, but yes the behavior will show in both dialogs, what we're currently running into you can reproduce from the following steps:

1. Create only a single Content Group on the Settings -> Group Configurations page, ex. "Content Group A"
2. In a new tab navigate to the Course Outline and open the unit Configuration option
3. On the Settings Group Configurations page, delete the content group you created, there should be none remaining and the You have not created any content groups yet. message should appear.
4. On the Course Outline tab with unit Configuration open, save the Unit configuration after selected the content group you had previously created and then deleted, and click save.
5. Refresh the page of the Course Outline tab, with a breakpoint on line 431 of contentstore/utils.py you'll only see the enrollment_track partitions from get_all_partitions_for_course.

The only way the previously Deleted Group appears is if there is at least one other content group created.

cahrens replied:
@jlajoie Thanks for investigating. If you can fix this bug in < 30 minutes, do so. Otherwise create a bug for it (with your excellent reproduction steps!) and move on.

cahrens commented on cms/templates/js/unit-access-editor.underscore:
```diff
@@ -0,0 +1,133 @@
+                        "deleted": true,
+                        "selected": true,
+                        "id": groupId,
+                        "name": "Deleted Group"
```
As an aside, you should never be adding non-translated strings to something that will be shown to a user.

cahrens commented on cms/templates/js/unit-access-editor.underscore:
```diff
@@ -0,0 +1,133 @@
+    <%- gettext('Unit Access') %>
+</h3>
+<div class="modal-section-content access-change">
+    <label for="partition-select" class="group-select-title"><%- gettext('Restrict access to:') %>
```
Nit: don't need "for" here because the nesting does the job. This means you can also get rid of the ID altogether and just use the user-partition-select class for selectors.

cahrens commented on cms/templates/js/unit-access-editor.underscore:
```diff
@@ -0,0 +1,133 @@
+        <div class="user-partition-group-checkboxes">
+            <% for (var i=0; i < selectablePartitions.length; i++) { %>
+                <div id="<%- selectablePartitions[i].id %>-checkboxes">
+                    <div class="partition-group-directions">
```
You need to use role="group" and aria-labelledby to link the directions. Make the HTML the same as in visibility_editor.html, as we took care there to handle accessibility concerns.

cahrens commented on cms/templates/js/unit-access-editor.underscore:
```diff
@@ -0,0 +1,133 @@
+            <% } %>
+        </div>
+    <% } else { %>
+        <%- gettext('You have not set up any content groups yet.') %>
```
As I mentioned before, I don't think this is the right message to show in this case. Please work with @roderickmorales.

cahrens commented on cms/templates/js/unit-access-editor.underscore:
```diff
@@ -0,0 +1,133 @@
+                                    <% // Checkbox is checked if the partition exists and the specific partition's group
+                                       // is in the xblock's group access, or if the group has been deleted. %>
+                                    <% if ((groupAccess[selectablePartitions[i].id]
+                                            && groupAccess[selectablePartitions[i].id].indexOf(selectablePartitions[i].groups[j].id) !== -1)
```
I'm confused about the data structure here. Is it not possible to pass through the data structure that already has the "selected" value (I'm looking here at the similar code in visibility_editor.html)?

cahrens commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -198,71 +205,129 @@ def _build_fixture(self, unit_state):
+                            [
+                                Group(1, 'Audit')
+                            ],
+                            scheme="cohort"
```
Scheme should not be cohort here (if we are trying to mimic enrollment tracks). You could use cohorts though if that is easier-- in that case, I just wouldn't make them "sound like" enrollment tracks.

cahrens commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -152,11 +156,12 @@ class PublishState(object):
         """ Represents the state of a unit """
 
-        def __init__(self, is_released, publish_state, is_locked):
+        def __init__(self, is_released, publish_state, is_locked, is_restricted):
```
Nit: you could make "is_restricted" have a default value (False) so that you don't have to update all the existing callers of the method.

cahrens commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -272,24 +337,87 @@ def test_unreleased_unpublished_changes_unlocked(self):
             None
         )
 
+    def test_unreleased_unpublished_unlocked_restricted(self):
```
Nit: you could use ddt to collapse these tests. Probably not worth it in this case because this test file is already following the pattern you are using, but something to look into for the future. Search for "ddt" inside the acceptance test folder to see examples.

cahrens commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -307,7 +435,8 @@ def _verify_unit_warning(self, unit_state, expected_status_message):
             self.assertFalse(section.has_status_message)
-            self.assertFalse(subsection.has_status_message)
+            if not self.GROUP_RESTRICTION_WARNING:
+                self.assertFalse(subsection.has_status_message)
```
Subsections should never have this warning, correct?

cahrens commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -492,6 +621,167 @@ def test_unchanged_release_date_is_not_saved(self):
+        self.assertFalse(unit.has_restricted_warning)
+        unit.set_unit_access('Content Groups', [self.content_group_a_id])
+        self.assertTrue(unit.has_restricted_warning)
+        unit.set_unit_access('Content Groups', [self.content_group_a_id])
```
"set_unit_access" is a confusing method name here because I would assume it always sets the access according to the input arguments. But apparently it is toggling?

cahrens commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -492,6 +621,167 @@ def test_unchanged_release_date_is_not_saved(self):
+        course_home_page.resume_course_from_header()
+        self.assertEqual(course_home_page.outline.num_units, 2)
+        course_home_page.visit()
+        course_home_page.preview.set_staff_view_mode("Learner in Test Group A")
```
There is a tension with bok choy tests-- normally, it is good to have separate tests to test different scenarios (as you have done here). But bok choy tests are notoriously slow. Therefore we tend to do more things in a single test point to avoid the setup time/cost.

I would put lines 738-741 at the end of the previous test point (and update the name accordingly) since the setup is all the same, and really we are testing what happens when you view as learners in different groups.

cahrens commented on common/test/acceptance/tests/studio/test_studio_outline.py:
```diff
@@ -492,6 +621,167 @@ def test_unchanged_release_date_is_not_saved(self):
+        course_home_page.resume_course_from_header()
+        self.assertEqual(course_home_page.outline.num_units, 2)
+        course_home_page.visit()
+        course_home_page.preview.set_staff_view_mode("Learner in Verified")
```
Same note-- I would collapse this into the previous test point.

cahrens commented:
@jlajoie and @bbaker6225 I'm done with this pass-- it's definitely getting better! Let me know when these comments have been addressed (and feel free to grab me if you have any questions).

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -633,6 +623,108 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+
+        getSelectedGroupAccess: function() {
+            var selectedEnrollmentTrackId = this.getSelectedEnrollmentTrackId();
+            this.getSelectedCheckboxesByDivId(selectedEnrollmentTrackId);
```
How does this method work as it has no return at the end of it?

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -633,6 +623,108 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+            'change .user-partition-select': function() {
+                this.hideCheckboxDivs();
+                this.showSelectedDiv(this.getSelectedEnrollmentTrackId());
+                this.getSelectedGroupAccess();
```
Why are you calling getSelectedGroupAccess here?

cahrens commented on cms/static/sass/elements/_modal-window.scss:
```diff
@@ -777,4 +780,44 @@
+
+  .edit-unit-access, .edit-staff-lock {
+    .modal-section-content {
+    @include font-size(16);
```
indenting

cahrens commented on cms/static/sass/elements/_modal-window.scss:
```diff
@@ -777,4 +780,44 @@
+      margin-bottom: ($baseline/4);
+
+      .user-partition-select {
+        font-size: inherit;
```
Is font-size: inherit needed here?

jlajoie replied:
We use inherit so that the children elements receive the same font size from the modal-section-content parent. Without the inherit we end up with a font size of 12 for the text inside of the modal.

cahrens commented on cms/static/sass/elements/_modal-window.scss:
```diff
@@ -777,4 +780,44 @@
+      }
+
+      font-size: inherit;
+      margin-left: ($baseline/4);
```
Anytime I see a margin-left (or right), I wonder what will happen with RTL. Have you tried it with RTL? There is an include for margin-left that will flip automatically for you-- @include margin-left()

jlajoie replied:
Oh cool, we hadn't tried with RTL, we can swap this over to use the include

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -57,11 +91,42 @@ if (staffOnlyMessage) {
+    } else if (numAllowedGroups >= 1 && xblockInfo.isVertical()) {
+        var selectedGroupsLabel = '';
+
+        for (var schemeId in groupAccessMap) {
```
I'm a bit confused here-- you are looping over all schemes? The selected groups should all be within a single scheme.

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -57,11 +91,42 @@ if (staffOnlyMessage) {
+                if (selectedGroupsLabel.length === 0) {
+                    selectedGroupsLabel = currentGroup;
+                } else {
+                    // Translators: This is building up a list of groups. It is marked for translation because of the
```
I don't think this string will get through to the translators because it is not directly above the gettext call. Read through all the comments about Translator comments in http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/internationalization/i18n.html#multi-line-strings.

In general, getting translator comments through in JavaScript files can be quite tricky. To actually be sure that it is working, you probably would need to generate translation files locally.

cahrens commented on cms/templates/js/unit-access-editor.underscore:
```diff
@@ -0,0 +1,46 @@
+                        <% for (var j = 0; j < userPartitions[i].groups.length; j++) { %>
+                            <div class="field partition-group-control">
+                                <input type="checkbox" id="content-group-<%- userPartitions[i].groups[j].id %>" value="<%- userPartitions[i].groups[j].id %>" class="input input-checkbox"
+                                    <%  if (userPartitions[i].groups[j].selected || userPartitions[i].groups[j].deleted) { %> checked="checked" <% } %> />
```
Interesting-- if a group is deleted, it isn't also marked as selected (since deleted groups are only returned if they are selected)? Looking at visibility_editor.html, it appears that you only need to check selected, even for deleted groups.

cahrens commented on cms/templates/js/staff-lock-editor.underscore:
```diff
@@ -7,26 +7,19 @@
+    <% } %>
 </div>
-</form>
+</form>
```
Nit: put a return at the end of the files (this one and the next).

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -4,6 +4,17 @@ var visibilityState = xblockInfo.get('visibility_state');
 var published = xblockInfo.get('published');
 var prereq = xblockInfo.get('prereq');
 var hasPartitionGroups = xblockInfo.get('has_partition_group_components');
+var userPartitions = xblockInfo.get('user_partitions');
```
@bbaker6225 As we talked about, try adding the return value from get_visibility_partition_info to xblockInfo and use that in here instead of reproducing it in this file. xblockInfo is already a dumping ground of information... I don't feel bad about adding more.

jlajoie commented:
jenkins run python

cahrens commented:
@sstack22 @catong @roderickmorales I have signed off on this PR-- do any of you need to review before we mark it "done"? Note that this is staying as a feature branch, so we can always iterate on it as we finish the rest of the stories.

catong commented:
Looks great to me; thanks! 👍

ssemenova commented on cms/templates/js/course-outline.underscore:
```diff
@@ -57,7 +60,16 @@ if (staffOnlyMessage) {
     }
 
-    if (hasPartitionGroups) {
+    if (selectedPartitionIndex !== -1 && !isNaN(selectedPartitionIndex) && xblockInfo.isVertical()) {
```
I don't really know javascript that well, but is !isNan(selectedPartitionIndex) the cleanest way to do this? Is selectedPartitionIndex ever below 0?

jlajoie replied:
My understanding is that user partition IDs shouldn't ever be below 0, the !isNaN is used mostly as a conditional check that the selectedPartitionIndex exists and wasn't parsed inappropriately, ex. ```
> !isNaN({})
false
> !isNaN(undefined)
false
> !isNaN(parseInt('a'))
false
```

cahrens commented:
FYI, you are now good to squash commits.

ssemenova commented:
jenkins run all

jzoldak commented:
@jlajoie @ssemenova @cahrens looks like this merge push us over the limit for XSSLinter violations on master.

```
Build failed running pavelib.quality.run_xsslint: XSSLinter Failed.
Too many javascript-jquery-append violations (78).
```

jlajoie commented:
@jzoldak is the easiest option for going about fixing this just to revert?
